### PR TITLE
Mouse selection support

### PR DESCRIPTION
**Pull request by sh7dm**
TODO:
- [x] Multiple selections created by mouse using modifier keys
- ~~Double-clicking to select words (will probably need to add extra vars to state for counting)~~
- ~~Select lines by dragging on the left side~~
- [x] Automatically switch view if clicked out of the current view
- ~~Pick items from menus using mouse (may be out of scope)~~

### CODE REVIEW

File: helix-term/src/ui/editor.rs
Comment (archseer) [inline helix-term/src/ui/editor.rs:669]: This doesn't really account for split views, graphemes or tabs.

You should use the raw coordinates and look them up on the view tree (for now you can probably just loop over `editor.tree.views()` and find the one with an overlapping `view.area`). Then there should be a complement method to https://github.com/helix-editor/helix/blob/a6aad3122d49ba07896ce047b5ae23da34eae803/helix-view/src/view.rs#L130-L162

`pos_at_screen_coords` that will give you the character position.

**Comment by archseer** [comment]
I think you can skip the double-clicking, it doesn't seem to be present on kakoune at least, not sure about vim.

File: helix-term/src/ui/editor.rs
Comment (sh7dm) [reply helix-term/src/ui/editor.rs]: Finished, thank you. Will add dragging on line numbers right now.

File: helix-view/src/view.rs
Comment (archseer) [inline helix-view/src/view.rs:190]: This is still not handling graphemes/visual width correctly, you need to fetch the current line, then traverse similar to https://github.com/helix-editor/helix/blob/ba70331b6b9b15d66fc8274b32464643bd9f2a97/helix-view/src/view.rs#L148-L155 until you reach the necessary column. The code in it's current state would treat a tab as four chars for example.

**Comment by cessen** [comment]
Really looking forward to this!

I agree that double-click is probably not needed.  If I want to select a bunch of words with the mouse, I'll probably just add cursors at each word with modifier-click, and then use a keyboard command to expand to word boundaries.  That way it's "click, click, click, hotkey" rather than "click-click, click-click, click-click".

Similarly for selecting a range of lines: I'd probably just drag the selection to encompass the lines I want, and then hit `X` to expand to line boundaries.  That also frees up clicking in the gutter for other things (not entirely sure what those would be, but maybe things related to debugging or version control).

**Review by archseer — CHANGES_REQUESTED** [review]

File: helix-term/src/ui/editor.rs
Comment (sh7dm) [reply helix-term/src/ui/editor.rs:776]: Well-noticed, doesn't seem to be necessary

File: helix-view/src/view.rs
Comment (sh7dm) [reply helix-view/src/view.rs:190]: Thank you, now I understood at which conditions to test. Will fix

**Comment by sh7dm** [comment]
> Similarly for selecting a range of lines: I'd probably just drag the selection to encompass the lines I want, and then hit X to expand to line boundaries. That also frees up clicking in the gutter for other things (not entirely sure what those would be, but maybe things related to debugging or version control).

You're right. If anyone wished that feature, right click can be used, but IMO that can be fully dropped really.

**Comment by sh7dm** [comment]
I've split out `verify_screen_coords`, but looks like there's no way to switch views from event handler. Looking for advice on how to do that the best.

**Comment by Omnikar** [comment]
Woah, I had no idea mouse integration in a terminal program was even possible.

**Comment by archseer** [comment]
> switch views from event handler

You can do this by setting `cx.editor.tree.focus = view.id`.

File: helix-view/src/view.rs
Comment (archseer) [inline helix-view/src/view.rs:227]: Why is this looping through the graphemes twice?

Comment (sh7dm) [reply helix-view/src/view.rs:227]: Actually, that's caused by using old code. Now that comparison can be dropped I think. Just stop iterating once target column is reached, if it can't be reached the loop will exit.

**Review by pickfire — APPROVED** [review]
For an initial implementation I think we can merge this step by step rather than all at once. Just left click is good enough as the first step. I think we can have them as separate PR?

**Comment by sh7dm** [comment]
Currently there's some work on view switching also, but that can be sent in another PR if it succeeds.

**Comment by sh7dm** [comment]
View switching is done locally. Should I push it here or do that later?

**Comment by archseer** [comment]
Yeah include it, I'd like to wait for at least @cessen's review too

**Comment by sh7dm** [comment]
Now two TODOs are left for another PR:
- Multiple selections created by mouse using modifier keys
- Pick items from menus using mouse (may be out of scope)

**Comment by archseer** [comment]
> * Multiple selections created by mouse using modifier keys

This shouldn't be too hard. If shift is pressed, instead of making a new selection clone the old one then use selection.push to add a new range. It will automatically become the new primary one, so the dragging code needs no changes.

I'd also like to see this behavior be configurable, if mouse = false in the config then the events should just not get enabled (I personally don't use the mouse and will disable it to reduce the amount of incoming crossterm events)

File: helix-term/src/ui/editor.rs
Comment (sudormrfbin) [inline helix-term/src/ui/editor.rs:799]: I think you can simply unwrap here since it's guaranteed by `verify_screen_coords` to be valid.

File: helix-term/src/ui/editor.rs
Comment (sudormrfbin) [inline helix-term/src/ui/editor.rs:825]: A match expression would be more ergonomic:

```suggestion
                let pos = match view.pos_at_screen_coords(doc, row as usize, column as usize) {
                    Some(pos) => pos,
                    None => return EventResult::Ignored,
                };
```

**Comment by sh7dm** [comment]
Maybe it would be nice to have a runtime command to change mouse behavior?

File: helix-term/src/ui/editor.rs
Comment (sh7dm) [reply helix-term/src/ui/editor.rs:799]: Well, true, thanks

File: helix-view/src/tree.rs
Comment (archseer) [inline helix-view/src/tree.rs:256]: This method is unnecessary, you can already access the id via view.id

File: helix-term/src/ui/editor.rs
Comment (archseer) [inline helix-term/src/ui/editor.rs:785]: Rather than for loop which will iterate over everything, use .find` since views_mut() is already an iterator.

File: helix-term/src/ui/editor.rs
Comment (archseer) [inline helix-term/src/ui/editor.rs:810]: Just call it id, no need for a separate variable name

File: helix-term/src/ui/editor.rs
Comment (archseer) [reply helix-term/src/ui/editor.rs:785]: Also rather than having a verify method and one that returns a position, you can use find_map

**Review by archseer — CHANGES_REQUESTED** [review]

File: helix-term/src/ui/editor.rs
Comment (sh7dm) [reply helix-term/src/ui/editor.rs]: Pushed separately with necessary changes.

Rust's `match` is wonderful.

File: helix-view/src/tree.rs
Comment (sh7dm) [reply helix-view/src/tree.rs:256]: All `Tree` fields apart from `focus` are private AFAIK:

`helix-view/src/tree.rs:7`
```rs
pub struct Tree {
    root: ViewId,
    // (container, index inside the container)
    pub focus: ViewId,
    // fullscreen: bool,
    area: Rect,

    nodes: HopSlotMap<ViewId, Node>,

    // used for traversals
    stack: Vec<(ViewId, Rect)>,
}
```

File: helix-term/src/application.rs
Comment (archseer) [inline helix-term/src/application.rs:452]: Let's just call it mouse and have it default to true (https://serde.rs/attr-default.html)

File: helix-view/src/tree.rs
Comment (archseer) [reply helix-view/src/tree.rs:256]: Why do you need access to nodes?

Comment (archseer) [reply helix-view/src/tree.rs:256]: The body inside MouseEventKind::Down can simply be something like:

```rust
let result = tree.views_mut().find_map(|(view, _focus)| {
  // no jumps
  let doc = &mut editor.documents[view.doc];
  view.pos_at_screen_coords(doc, row, column).map(|coords| (view.id, coords))
})

if let Some((view_id, coords)) = result {
   ... set_selection etc
}
```

File: helix-view/src/view.rs
Comment (archseer) [inline helix-view/src/view.rs:217]: Needs to be a saturating sub probably otherwise clicking on (0, 0) will panic, also why the `- 1`?

Comment (sh7dm) [reply helix-view/src/view.rs:217]: User clicks after the line end, all the length gets counted and so overflowed to the next line. No panic happens with (0, 0)

File: helix-view/src/tree.rs
Comment (sh7dm) [reply helix-view/src/tree.rs:256]: I wasn't able to use `views_mut` because I later need the view id to later focus it in the editor
```rs
if let Some(id) = id {
    cx.editor.tree.focus = id;
}
```

Comment (archseer) [reply helix-view/src/tree.rs:256]: `view.id` is `Copy`, so you just need to dereference it so it gets copied and that way `let result` isn't bound to the tree.

File: helix-term/src/ui/editor.rs
Comment (sh7dm) [reply helix-term/src/ui/editor.rs:785]: You mean `pos_at_screen_coords` should not validate its input and thus panic? Well, with doc comment stating that should be absolutely fine.

File: helix-view/src/tree.rs
Comment (sh7dm) [reply helix-view/src/tree.rs:256]: Will fix, thank you!

File: helix-term/src/ui/editor.rs
Comment (archseer) [reply helix-term/src/ui/editor.rs:785]: I meant https://github.com/helix-editor/helix/pull/509#discussion_r678282400 <- find_map already does the verification

File: helix-view/src/view.rs
Comment (archseer) [reply helix-view/src/view.rs:217]: I mean if I click on the top left first character of the first line. `pos` there will be 0 I think?

File: helix-term/src/ui/editor.rs
Comment (sh7dm) [inline helix-term/src/ui/editor.rs:833]: Notice: this is incorrect, I'm working on a fix. This way won't remove old selection while dragged so selection will only grow following the cursor.

File: helix-view/src/view.rs
Comment (sh7dm) [reply helix-view/src/view.rs:217]: Looks like not, since nothing crashes. Anyway, it'd be nice to count whole line lenght - 1 with the loop iterating over graphemes.

File: helix-view/src/view.rs
Comment (cessen) [inline helix-view/src/view.rs:186]: Instead of returning `None` when the cursor is outside of the text, it might be more useful to return the "closest" valid position in the text.  For example, even if the mouse is off the end of a line, return the position of the last character on that line.  This would make it easier to e.g. quickly drag-select a bunch of lines without having to be too precise.

I suspect that's also the behavior we would want if we eventually use this function elsewhere, too.

Comment (cessen) [reply helix-view/src/view.rs:186]: Also, we'll definitely want unit tests on this.  It might be worth breaking out a sub-function that does all of the actual logic, but takes a `RopeSlice` instead of a `Document`, so that unit tests are easier to write.

File: helix-view/src/view.rs
Comment (cessen) [inline helix-view/src/view.rs:188]: I think we need to ensure that `line_number < text.len_lines()` here.  Imagine a document with only a few lines, and the user clicks in an area within the view but below the valid text lines.

I think what we probably want to do is check, and if the line number is `>= text.len_lines()` then immediately return the final text position (`text.len_chars() - 1`).

File: helix-view/src/view.rs
Comment (cessen) [inline helix-view/src/view.rs:210]: This needs to be `pos += grapheme.chars().count();`.

File: helix-view/src/view.rs
Comment (cessen) [inline helix-view/src/view.rs:206]: This should just be `else`.  We shouldn't be ignoring newline characters here.

(And a note for the future: when you do need to handle them specially, you need to handle *all* line ending characters, not just `\n`.  The function `helix_core::line_ending::str_is_line_ending()` will do all the needed checks for you.)

File: helix-view/src/view.rs
Comment (cessen) [inline helix-view/src/view.rs:201]: This can be `>=`, and then we don't need the `- 1` on line 217 below.

File: helix-view/src/view.rs
Comment (cessen) [reply helix-view/src/view.rs]: Along with the comments I made above, I think what we want instead of `Some(pos - 1)` here is `Some(pos.min(line_end_char_index(text.slice(..), line_number)))`.  Basically, just ensuring that it doesn't go off the end of the line.  (You can find `line_end_char_index()` in `helix_core::line_ending`.)

Getting code like this right can be subtle, so I may have missed some things or made a mistake in my suggestions.  I definitely would feel more comfortable about this with unit tests, to make sure we're getting things right.

File: helix-view/src/view.rs
Comment (sh7dm) [reply helix-view/src/view.rs:201]: Yes. Both were changed because full line length (click after line end) needs to be reduced by 1.

Comment (cessen) [reply helix-view/src/view.rs:201]: Yeah, both approaches will work.  I think the code ends up a little clearer with `>=` and handling the line ending explicitly, but I don't feel too strongly about it.

Having said that, the `Some(pos - 1)` at the end is still incorrect in your version.  It needs to work in terms of graphemes, so it should be `Some(prev_grapheme_boundary(text, pos))`.

(Your version will also break on the last line of the file after #376 is merged, since there's no guaranteed line break on the last line with that PR.  So we'll need to change it after that anyway.  But I don't think we need to block this PR for that.)

**Comment by Omnikar** [comment]
Is it intended for drag-selection to create multiple cursors?

**Comment by pickfire** [comment]
> Is it intended for drag-selection to create multiple cursors?

Not by default but probably possible with some combinations like ctrl. Like ctrl + drag a few stuff.

**Comment by sh7dm** [comment]
Alt+Click makes a new cursor. Drag is a bit broken atm, see https://github.com/helix-editor/helix/pull/509#discussion_r678500958

File: helix-term/src/ui/editor.rs
Comment (archseer) [inline helix-term/src/ui/editor.rs:783]: Call `verify_screen_coords` inside `pos_at_screen_coords` and return an `Option` from pos_at_screen_coords. You can actually inline it completely and remove the function. Then all you need to do here is 

```rust
find_map(|(view, _focus)| {
//...;
view.pos_at_screen_coords(...).map(|coords| (coords, view.id))
})
```

File: helix-term/src/ui/editor.rs
Comment (archseer) [inline helix-term/src/ui/editor.rs:797]: I don't think this is necessary? Just refer to `editor.tree`

File: helix-term/src/ui/editor.rs
Comment (archseer) [reply helix-term/src/ui/editor.rs:836]: Instead of pushing you want just edit the primary selection range.

**Review by archseer — CHANGES_REQUESTED** [review]
Looking better now, a couple more changes

File: helix-term/src/ui/editor.rs
Comment (sh7dm) [reply helix-term/src/ui/editor.rs:783]: We've decided (cannot reference conversation, will try to find later) that it is not necessary to check inside it as we already validate view while searching for it.

File: helix-term/src/ui/editor.rs
Comment (sh7dm) [reply helix-term/src/ui/editor.rs:797]: I did split it out because of multiple references. Can refer directly if needed by code style.

File: helix-term/src/ui/editor.rs
Comment (sh7dm) [reply helix-term/src/ui/editor.rs:783]: Although if they're united it can be better.

File: helix-view/src/view.rs
Comment (cessen) [inline helix-view/src/view.rs:200]: Sorry for the run-around on this: now that #376 has been merged, this should just be `Some(text.len_chars())`, because to-the-right of the last character is now the last valid document position.

Comment (sh7dm) [reply helix-view/src/view.rs:200]: Thanks, I did not repeat test with short file after branch update.

**Review by cessen — APPROVED** [review]
Looks good to me!  I also tested locally, and everything seems to work great.

(I think we could use some more thorough tests, including e.g. CJK characters, graphemes, etc. but I can make a PR for that later myself.  The basics are there.)

Note: the lint failures in the recent CI run are due to #525, not this PR.

**Comment by cessen** [comment]
@sh7dm
If you rebase on current master now, the lints should pass.

**Comment by archseer** [comment]
Great work! I followed up with some refactoring (https://github.com/helix-editor/helix/commit/6bb744aeacb56846c3bb50338e2bb81328e4e396, https://github.com/helix-editor/helix/commit/62eb8c6b407c61d21dbee83c4b4e78c4e6cb87c3)

**Comment by sh7dm** [comment]
Huge thanks to all the reviewers for the great help!

**Comment by cessen** [comment]
Great work, @sh7dm!  Thanks for taking the time and effort for this!

**Pull request merged**
